rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/FileDownloadRequestChain.java

```diff
@@ -156,39 +152,33 @@ public Integer call() throws IOException
   private int copyIntoCache(FSDataInputStream inputStream, FileChannel fileChannel, int length,
                             long cacheReadStart) throws IOException
   {
-    byte[] buffer = null;
+    long start = System.nanoTime();
+    byte[] buffer = new byte[length];
+    log.debug(String.format("Copying data of file %s of length %d from position %d", remotePath, length, cacheReadStart));
     int nread = 0;
-    try {
-      long start = System.nanoTime();
-      buffer = new byte[length];
-      log.debug("Copying data of file " + remotePath + " of length " + length + " from offset " + cacheReadStart);
-      while (nread < length) {
-        int nbytes = inputStream.read(buffer, nread, length - nread);
-        if (nbytes < 0) {
-          break;
-        }
-        nread += nbytes;
-      }
-
-      int leftToWrite = length;
-      int writtenSoFar = 0;
-
-      while (leftToWrite > 0) {
-        int writeInThisCycle = Math.min(leftToWrite, directBuffer.capacity());
-        directBuffer.clear();
-        directBuffer.put(buffer, writtenSoFar, writeInThisCycle);
-        directBuffer.flip();
-        int nwrite = fileChannel.write(directBuffer, cacheReadStart + writtenSoFar);
-        directBuffer.compact();
-        writtenSoFar += nwrite;
-        leftToWrite -= nwrite;
+    while (nread < length) {
+      int nbytes = inputStream.read(cacheReadStart + nread, buffer, nread, length - nread);
```

Comment:
There seems to be a bug in existing code where we are allowing to read < length but later try to write to file `length` bytes. But we would not hit it given we already correct the ReadRequest backendReadEnd if it is going beyond eof.
We can replace this while loop with readFully call.

Reply:
Are we assured here that `length` won't go beyond EOF ? I had avoided readFully here because I wasn't sure whether `readRequest.getBackendReadLength()` takes into account size of file while aligning reads to block boundaries.
I think `int leftToWrite = nread;` would fix the bug you've mentioned ?

Reply:
Yes, that or using readFully (that will error out incase we failed to correct the end of read) should do

Reply:
Used readFully here